libraries/esp8266/examples/LowPowerDemo/LowPowerDemo.ino

```diff
@@ -0,0 +1,338 @@
+#define LED 5  // D1/GPIO5 external LED for modules with built-in LEDs so it doesn't add to the amperage
+//#define LED 2  // D4/GPIO2 LED for ESP-01,07 modules; D4 is LED_BUILTIN on most other modules
+// you can use LED_BUILTIN, but it adds to the measured amperage by 0.3mA to 6mA.
+
+ADC_MODE(ADC_VCC);  // allows you to monitor the internal VCC level; it varies with WiFi load
+// don't connect anything to the analog input pin(s)!
+
+// enter your WiFi configuration below
+const char* AP_SSID = "SSID";  // your router's SSID here
+const char* AP_PASS = "password";  // your router's password here
+IPAddress staticIP(0, 0, 0, 0); // parameters below are for your static IP address, if used
+IPAddress gateway(0, 0, 0, 0);
+IPAddress subnet(0, 0, 0, 0);
+IPAddress dns1(0, 0, 0, 0);
+IPAddress dns2(0, 0, 0, 0);
+uint32_t wifiTimeout = 25E3;  // 25 second timeout on the WiFi connection
+
+// This structure will be stored in RTC memory to remember the reset count (number of Deep Sleeps).
+// First field is CRC32, which is calculated based on the rest of the structure contents.
+// Any fields can go after the CRC32.  The structure must be 4-byte aligned.
+struct {
+  uint32_t crc32;
+  byte data[4];  // the last byte stores the reset count
+} rtcData;
+
+byte resetCount = 0;  // keeps track of the number of Deep Sleep tests / resets
+String resetCause = "";
+
+const uint32_t blinkDelay = 100; // fast blink rate for the LED when waiting for the user
+const uint32_t longDelay = 350;  // longer delay() for the two AUTOMATIC modes
+esp8266::polledTimeout::periodicFastMs blinkLED(blinkDelay);
+// use fully qualified type and avoid importing all ::esp8266 namespace to the global namespace
+
+void wakeupCallback() {  // unlike ISRs, you can do a print() from a callback function
+  Serial.println(F("Woke from Forced Light Sleep - this is the callback"));
+}
+
+void setup() {
+  pinMode(LED, OUTPUT);  // activity and status indicator
+  digitalWrite(LED, LOW);  // turn on the LED
+  pinMode(WAKE_UP_PIN, INPUT_PULLUP);  // polled to advance tests, interrupt for Forced Light Sleep
+  Serial.begin(115200);
+  Serial.print(F("\nReset reason = "));
+  String resetCause = ESP.getResetReason();
+  Serial.println(resetCause);
+  if ((resetCause == "External System") || (resetCause == "Power on")) {
+    Serial.println(F("I'm awake and starting the Low Power tests"));
+    resetCount = 5;
+    updateRTC();  // if external reset, wipe the RTC memory and start all over
+  }
+
+  // Read struct from RTC memory
+  if (ESP.rtcUserMemoryRead(64, (uint32_t*) &rtcData, sizeof(rtcData))) {
+    uint32_t crcOfData = crc32((uint8_t*) &rtcData.data[0], sizeof(rtcData.data));
+    if (crcOfData != rtcData.crc32) {  // if the CRC is invalid
+      resetCount = 0;  // set first test run since power on or external reset
+    } else {
+      resetCount = rtcData.data[3];  // read the previous reset count
+    }
+  }
+}  // end of setup()
+
+void loop() {
```

Comment:
This loop body is too long, and therefore hard to follow. I suggest factoring out each test into its own function. For example:
```cpp
void loop() {
  if(resetCount == 0) {
    runTest1();
    runTest2();
    runTest3();
    runTest4();
    runTest5();
    runTest6();
  }
  if(resetCount < 4) {
    initWiFi();
  }
  if(resetCount == 1) {
    runTest7();
  } else if(resetcount == 2) {
    runTest8();
  } else if (resetCount == 3 {
    runTest9();
  } else if (resetCount == 4) {
    resetTests();
  }
} //end loop

void runTest1() {
   // 1st test - running with WiFi unconfigured, reads ~67 mA minimum
    Serial.println(F("\n1st test - running with WiFi unconfigured"));
    float volts = ESP.getVcc();
    Serial.printf("The internal VCC reads %1.3f volts\n", volts / 1000);
    Serial.println(F("press the button to continue"));
    waitPushbutton(false, blinkDelay);
}

...

void resetTests() {
  resetCount = rtcData.writeToRTC(5);  // start all over, do an ESP.restart to insure it's a clean state
  float volts = ESP.getVcc();
  Serial.printf("The internal VCC reads %1.3f volts\n", volts / 1000);
  Serial.println(F("\nTests completed, in RF_DISABLED, press the button to do an ESP.restart()"));
  waitPushbutton(false, 1000);
  ESP.restart();
}
```

Reply:
Done, committed, pushed, passed Travis